ossfuzz_py/core/benchmark_manager.py

```diff
@@ -0,0 +1,702 @@
+import json
+import logging
+import yaml
+from collections import OrderedDict
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Dict, List, Any, Optional
+
+from ossfuzz_py.core.data_models import FileType
+from ossfuzz_py.core.errors import BenchmarkError, BenchmarkValidationError
+from ossfuzz_py.utils.file_utils import FileUtils
+
+# Configure module logger
+logger = logging.getLogger('ossfuzz_sdk.benchmark_manager')
+
+
+@dataclass(frozen=True)
+class Benchmark:
+    """
+    Immutable data class that stores metadata for a single benchmark/function-under-test.
+
+    This is a pure data holder containing all benchmark metadata for fuzzing,
+    including function signature information, project details, and optional
+    configuration flags. All business logic is handled by BenchmarkManager.
+
+    Fields are identical to the original benchmark.py implementation to ensure
+    compatibility with existing benchmark files.
+    """
+
+    # Core identification fields (required)
+    id: str = field(metadata={"description": "Unique benchmark identifier"})
+    project: str = field(metadata={"description": "OSS-Fuzz project name"})
+    language: str = field(metadata={"description": "Primary language (C/C++/Rust/...)"})
+    function_signature: str = field(metadata={"description": "Full function signature string"})
+    function_name: str = field(metadata={"description": "Parsed function name"})
+    return_type: str = field(metadata={"description": "Function return type"})
+    target_path: str = field(metadata={"description": "Path to header/source file"})
+
+    # Optional fields with defaults
+    params: List[Dict[str, str]] = field(
+        default_factory=list,
+        metadata={"description": "List of {name, type} parameter dicts"}
+    )
+    target_name: Optional[str] = field(
+        default=None,
+        metadata={"description": "Preferred target name (optional)"}
+    )
+    use_project_examples: bool = field(
+        default=True,
+        metadata={"description": "Whether to use project examples"}
+    )
+    use_context: bool = field(
+        default=False,
+        metadata={"description": "Whether to use context information"}
+    )
+    cppify_headers: bool = field(
+        default=False,
+        metadata={"description": "Whether to cppify headers"}
+    )
+    commit: Optional[str] = field(
+        default=None,
+        metadata={"description": "OSS-Fuzz project commit hash"}
+    )
+    test_file_path: str = field(
+        default='',
+        metadata={"description": "Path to test file (for test-based benchmarks)"}
+    )
+    function_dict: Optional[Dict[str, Any]] = field(
+        default=None,
+        metadata={"description": "Original function dictionary from YAML"}
+    )
+
+    def __hash__(self) -> int:
+        """Hash based on benchmark id for use in sets and as dict keys."""
+        return hash(self.id)
+
+    @property
+    def benchmark_id(self) -> str:
```

Comment:
nit: Will this cause stuttering? E.g., `benchmark.benchmark_id`.
Would `benchmark.id` be better?